src/psyclone/psyir/backend/fortran.py

```diff
@@ -540,18 +620,20 @@ def binaryoperation_node(self, node):
         :rtype: str
 
         '''
-        # reverse the fortran2psyir mapping to make a psyir2fortran
-        # mapping
-        mapping = _reverse_map(Fparser2Reader.binary_operators)
         lhs = self._visit(node.children[0])
         rhs = self._visit(node.children[1])
         try:
-            oper = mapping[node.operator]
-            # This is a binary operation
-            if oper.upper() in FORTRAN_INTRINSICS:
+            fort_oper = get_fortran_operator(node.operator)
+            if is_fortran_intrinsic(fort_oper):
                 # This is a binary intrinsic function.
-                return "{0}({1}, {2})".format(oper.upper(), lhs, rhs)
-            return "{0} {1} {2}".format(lhs, oper, rhs)
+                return "{0}({1}, {2})".format(fort_oper, lhs, rhs)
+            if isinstance(node.parent, Operation):
+                parent_fort_oper = get_fortran_operator(node.parent.operator)
+                if not is_fortran_intrinsic(parent_fort_oper) and \
+                   precedence(fort_oper) < precedence(parent_fort_oper):
```

Comment:
There is still a subtle error here since floating point operation are not associative we still need to care about the order of operators with the same precedence level.
For instance, at the moment:
`a + (b + c)` would be rewritten as `a + b + c`, which for FP is not right.

You can solve it using `<=` instead of `<`, but this will put unnecessary parenthesis which you are trying to avoid. A more targeted solution is to find out if this node is the lhs or rhs of the parent operation and just put the paranthesis if it is the rhs.
This should work assuming all equivalent operations have left-to-right precedende in Fortran, which I think is the case.

Reply:
Yes, well spotted. I've not seen code that does this before, but we should retain the original ordering. I went for the targeted solution.